### PR TITLE
chore: upgrade to TS 4.7, compile with NodeNext

### PR DESCRIPTION
**Pull request by Josh-Cena**
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [x] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.
- [ ] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #0000) and the maintainers have approved on my working plan.

## Motivation

This is still a pain in the ass, but at least better than nothing.

I had to add ambient declarations for a few packages (namely, `eta` and `webpackbar`) that have incompatible package.json manifest formats. I hope they can be fixed soon.

I also corrected the format of our own manifests to make sure they can be consumed by external projects compiled with `NodeNext`. I'm already having pain with this in my own projects.

This can be a step towards #6520.

### Test links

Deploy preview: https://deploy-preview-_____--docusaurus-2.netlify.app/

### CODE REVIEW

**Comment by Josh-Cena** [comment]
Holding off till new version of vfile is released with better type declaration

File: packages/docusaurus/src/commands/swizzle/actions.ts
Comment (Josh-Cena) [inline packages/docusaurus/src/commands/swizzle/actions.ts:131]: Ideally revert before merge—depends on if DT is open to fixing this.

Comment (slorber) [reply packages/docusaurus/src/commands/swizzle/actions.ts]: cf Discord

We'd rather export our own `ComponentProps` or `WrapperProps` type instead of inlining these more complex types

https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/60766

**Review by slorber — APPROVED** [review]
Great 👍 almost LGTM

File: packages/docusaurus-theme-classic/src/theme/CodeBlock/CopyButton/index.tsx
Comment (Josh-Cena) [inline packages/docusaurus-theme-classic/src/theme/CodeBlock/CopyButton/index.tsx:11]: This is specifically because Webpack resolution is neither `node` nor `NodeNext`. Webpack allows `exports`, importing ESM in CJS, importing CJS in ESM, using both `require` and `import` in one file, importing `.ts` extension, importing `.js` extension... everything.

https://twitter.com/atcb/status/1536831060960419841

File: packages/docusaurus-plugin-content-docs/src/client/index.ts
Comment (slorber) [reply packages/docusaurus-plugin-content-docs/src/client/index.ts:88]: I'd rather keep this `| undefined` for now, just for us, considering those client APIs are not documented

**Comment by slorber** [comment]
maybe the explicit pluginId is a bit overkill 🤷‍♂️ we'll see later if it's worth keeping

**Pull request merged**
